README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# scvitools

<!-- badges: start -->
<!-- badges: end -->

The scvitools R package is a wrap of scvi-tools python package which try
to simply the integration of scvi-tools with Seurat.

## Installation

You can install the development version of scvitools from
[GitHub](https://github.com/) with:

``` r
# install.packages("devtools")
devtools::install_github("EmanuelSoda/scvi-tools-R")
```

## Example

This is a basic example which shows you how to solve a common problem:

``` r
library(scvitools)
```

## Code of Conduct

Please note that the scvitools project is released with a [Contributor
Code of
Conduct](https://emanuelsoda.github.io/scvi-tools-R/CODE_OF_CONDUCT.html).
By contributing to this project, you agree to abide by its terms.


NEWS.md
# scvitools 0.1.0

* Added a `NEWS.md` file to track changes to the package.


tests/testthat.R
# This file is part of the standard setup for testthat.
# It is recommended that you do not modify it.
#
# Where should you do additional test configuration?
# Learn more about the roles of various files in:
# * https://r-pkgs.org/tests.html
# * https://testthat.r-lib.org/reference/test_package.html#special-files

library(testthat)
library(scvitools)

test_check("scvitools")


tests/testthat/test-scvi.R
test_that("test if we can create and train scvi model using CreateSCVI", {
    data("object", package = "scvitools")
    pbmc <- object
    model <- CreateSCVI(pbmc, max_epochs = 1)
    testthat::expect_true(model$is_trained)
})

test_that("test if we can create scvi model using CreateSCVI", {
    data("object", package = "scvitools")
    pbmc <- object
    model <- CreateSCVI(pbmc, max_epochs = 1, train = FALSE)
    testthat::expect_false(model$is_trained)
})

test_that("test if we can run RunSCVI", {
    data("object", package = "scvitools")
    pbmc <- object
    model <- CreateSCVI(pbmc, max_epochs = 1)
    pbmc <- RunSCVI(pbmc, scvi_model = model)
    testthat::expect_equal(Seurat::Reductions(object = pbmc), "scvi")
})


R/utils.R
#' @importFrom memoise memoise
.onAttach <- function(libname, pkgname) {
    scvi <<- memoise::memoise(load_scvi)
    sc <<- memoise::memoise(load_sc)
    #sceasy <<- memoise::memoise(load_sceasy)
}



load_sceasy <- function() {
    tryCatch({
        find.package("sceasy")
    },
    error = function(e){
        message(paste0("It seams that sceasy is not installed. \n",
                       "You can install it using: \n",
                       'devtools::install_github("cellgeni/sceasy")'))
    }
    )
}

#' Load scvi-tools Python Package
#' @importFrom reticulate py_discover_config import
load_scvi <- function() {
    py_config <- try(
        reticulate::py_discover_config(required_module = "scvi")
    )

    delay_load = list(on_load = check_pyscvi_version #,
                      #on_error = failed_pyphate_import
                      )
    # load
    scvi <- try(
        reticulate::import("scvi", delay_load = delay_load)
    ) #delay_load = delay_load

    return(scvi)
}


#' Check that the current scvi-tools version in Python is up to date.
#'
#' @importFrom utils packageVersion
#' @importFrom reticulate py_config
#' @export
check_pyscvi_version <- function() {
    pyversion <- strsplit(scvi()$`__version__`, '\\.')[[1]]
    rversion <- strsplit(as.character(packageVersion("scvi")), '\\.')[[1]]
    major_version <- as.integer(rversion[1])
    minor_version <- as.integer(rversion[2])
    if (as.integer(pyversion[1]) < major_version) {
        warning(paste0("Python scvi version ", scvi()$`__version__`,
                       " is out of date (recommended: ",
                       major_version, ".", minor_version,
                       "). Please update with pip ",
                       "(e.g. ", reticulate::py_config()$python,
                       " -m pip install --upgrade scvi-tools)"))
        return(FALSE)
    } else if (as.integer(pyversion[2]) < minor_version) {
        warning(paste0("Python scvi version ", pyphate()$`__version__`,
                       " is out of date (recommended: ",
                       major_version, ".", minor_version,
                       "). Consider updating with pip ",
                       "(e.g. ", reticulate::py_config()$python,
                       " -m pip install --upgrade scvi-tools)."))
        return(FALSE)
    }
    return(TRUE)
}

#' Install scvi-tools Python Package
#'
#' Install scvi-tools Python package into a virtualenv or conda env.
#'
#' On Linux and OS X the "virtualenv" method will be used by default
#' ("conda" will be used if virtualenv isn't available). On Windows,
#' the "conda" method is always used.
#'
#' @param envname Name of environment to install packages into
#' @param method Installation method. By default, "auto" automatically finds
#' a method that will work in the local environment. Change the default to
#' force a specific installation method. Note that the "virtualenv" method
#' is not available on Windows.
#' @param conda Path to conda executable (or "auto" to find conda using the PATH
#'  and other conventional install locations).
#' @param pip Install from pip, if possible.
#' @param ... Additional arguments passed to conda_install() or
#' virtualenv_install().
#' @importFrom reticulate py_install
#'
#' @export
install_scvi <- function(envname = "r-reticulate", method = "auto",
                          conda = "auto", pip=TRUE, ...) {
    tryCatch({
        message(paste0("Attempting to install scvi-tools ",
                       "Python package with reticulate"))
        reticulate::py_install("scvi-tools",
                               envname = envname, method = method,
                               conda = conda, pip=pip, ...
        )
        message("Install complete. Please restart R and try again.")
    },
    error = function(e) {
        stop(paste0(
            "Cannot locate scvi-tools Python package, ",
            "please install through pip ",
            "(e.g. ", reticulate::py_config()$python,
            " -m pip install --user scvi-tools) and then restart R."
        ))
    }
    )
}



#' Load scanpy Python Package
load_sc <- function() {
    py_config <- try(
        reticulate::py_discover_config(required_module = "sc")
    )

    delay_load = list(on_load = check_pysc_version #,
                      #on_error = failed_pyphate_import
    )
    # load
    sc <- try(
        reticulate::import("scanpy", delay_load = delay_load)
    ) #delay_load = delay_load

    return(sc)
}


#' Check that the current scvi-tools version in Python is up to date.
#'
#' @importFrom utils packageVersion
#' @export
check_pysc_version <- function() {
    pyversion <- strsplit(sc()$`__version__`, '\\.')[[1]]
    rversion <- strsplit(as.character(packageVersion("scanpy")), '\\.')[[1]]
    major_version <- as.integer(rversion[1])
    minor_version <- as.integer(rversion[2])
    if (as.integer(pyversion[1]) < major_version) {
        warning(paste0("Python scanpy version ", sc()$`__version__`,
                       " is out of date (recommended: ",
                       major_version, ".", minor_version,
                       "). Please update with pip ",
                       "(e.g. ", reticulate::py_config()$python,
                       " -m pip install --upgrade scanpy)"))
        return(FALSE)
    } else if (as.integer(pyversion[2]) < minor_version) {
        warning(paste0("Python scanpy version ", pyphate()$`__version__`,
                       " is out of date (recommended: ",
                       major_version, ".", minor_version,
                       "). Consider updating with pip ",
                       "(e.g. ", reticulate::py_config()$python,
                       " -m pip install --upgrade scanpy)."))
        return(FALSE)
    }
    return(TRUE)
}

#' Install scanpy Python Package
#'
#' Install sscanpy Python package into a virtualenv or conda env.
#'
#' On Linux and OS X the "virtualenv" method will be used by default
#' ("conda" will be used if virtualenv isn't available). On Windows,
#' the "conda" method is always used.
#'
#' @param envname Name of environment to install packages into
#' @param method Installation method. By default, "auto" automatically finds
#' a method that will work in the local environment. Change the default to
#' force a specific installation method. Note that the "virtualenv" method
#' is not available on Windows.
#' @param conda Path to conda executable (or "auto" to find conda using the PATH
#'  and other conventional install locations).
#' @param pip Install from pip, if possible.
#' @param ... Additional arguments passed to conda_install() or
#' virtualenv_install().
#'
#' @export
install_scvi <- function(envname = "r-reticulate", method = "auto",
                         conda = "auto", pip=TRUE, ...) {
    tryCatch({
        message(paste0("Attempting to install scanpy ",
                       "Python package with reticulate"))
        reticulate::py_install("scanpy",
                               envname = envname, method = method,
                               conda = conda, pip=pip, ...
        )
        message("Install complete. Please restart R and try again.")
    },
    error = function(e) {
        stop(paste0(
            "Cannot locate scanpy Python package, ",
            "please install through pip ",
            "(e.g. ", reticulate::py_config()$python,
            " -m pip install --user scanpy) and then restart R."
        ))
    }
    )
}



RunSCVI.R
#' @title Run che scvi function
#' @description
#'              \code{\link{}}
#'
#'
#' @param object The ScreenR object obtained using the
#'                       \code{\link{create_screenr_object}}


#' @concept Run
#' @export
#' @rdname RunSCVI
#' @examples
#' object <- get0("object", envir = asNamespace("ScreenR"))
#' metrics <- compute_metrics(object,
#'     control = "TRT",
#'     treatment = "Time3", day = "Time3"
#' )
#' head(metrics)
RunSCVI <- function(object, ...) {
    UseMethod("RunSCVI")
}

#' @rdname RunSCVI
#' @param max_epochs Number of passes through the dataset.
#'       If NULL, defaults to min(round(c(round((20000 / n_cells) * 400), 400)))
#' @param early_stopping Perform early stopping.
#' @param use_gpu Use default GPU if available (if NULL or TRUE),
#'                or index of GPU to use (if int),
#'                or name of GPU (if string), or use CPU (if FALSE).
#' @return Return a Seurat object with the computed embedding
#'         with all the measure computed. For downstream Seurat analyses,
#'         use reduction='scvi'.
#' @export
RunSCVI.Seurat <- function(
        object,
        main_layer = "counts",
        max_epochs = NULL,
        use_gpu = NULL,
        early_stopping = FALSE

) {
    # if (!requireNamespace('Seurat')) {
    #     stop("Running SCVI on a Seurat object requires Seurat")
    # }

    assay.use <- Seurat::DefaultAssay(object)

    # creating the scanpy data
    adata <- sceasy::convertFormat(object, from = "seurat",
                                   to = "anndata",
                                   main_layer = main_layer,
                                   drop_single_values = FALSE)

    # setting up the scvi model
    scvi$model$SCVI$setup_anndata(adata)

    model <- scvi$model$SCVI(adata)

    # Tansform the R inpti into python input
    if (is.null(max_epochs)) {
        max_epochs <-  r_to_py(max_epochs)
    } else if(is.numeric(max_epochs)) {
        max_epochs <-  as.integer(max_epochs)
    }
    model$train(max_epochs = max_epochs,
                use_gpu = r_to_py(use_gpu),
                early_stopping = r_to_py(early_stopping))

    latent <- model$get_latent_representation()


    latent <- as.matrix(latent)
    rownames(latent) <- colnames(object)
    object[["scvi"]] <-
        CreateDimReducObject(embeddings = latent,
                             key = "scvi_",
                             assay = DefaultAssay(object))

    return(object)
}


R/scvi_model.R
#' Create and train SCVI model
#'
#' This function create and train a scvi model.
#' @param max_epochs Number of passes through the dataset.
#'       If NULL, defaults to min(round(c(round((20000 / n_cells) * 400), 400)))
#' @param early_stopping Perform early stopping.
#' @param use_gpu Use default GPU if available (if NULL or TRUE),
#'                or index of GPU to use (if int),
#'                or name of GPU (if string), or use CPU (if FALSE).
#' @param train If TRUE the model will be created and trained.
#' @concept Create scvi-tools model
#' @importFrom reticulate r_to_py
#' @rdname CreateSCVI
#' @export
CreateSCVI <- function(object, ...) {
    UseMethod("CreateSCVI")
}

#' @rdname CreateSCVI
#' @return Return a trained SCVI model.
#' @export
CreateSCVI.Seurat <- function(
        object,
        main_layer = "counts",
        max_epochs = NULL,
        use_gpu = NULL,
        train = TRUE,
        early_stopping = FALSE
) {
    # if (!requireNamespace('Seurat')) {
    #     stop("Running SCVI on a Seurat object requires Seurat")
    # }

    # creating the scanpy data
    adata <- sceasy::convertFormat(object, from = "seurat",
                                   to = "anndata",
                                   main_layer = main_layer,
                                   drop_single_values = FALSE)

    # setting up the scvi model
    scvi()$model$SCVI$setup_anndata(adata)

    model <- scvi()$model$SCVI(adata)

    if (train == TRUE) {
        model$train(max_epochs = is.integer(max_epochs),
                    use_gpu = reticulate::r_to_py(use_gpu),
                    early_stopping = reticulate::r_to_py(early_stopping))
    }

    return(model)
}

#' scvi-tools single cell
#'
#' Run dimentionality reduction usinc SCVI model
#' @param scvi_model A scvi_model
#' @concept Run scvi-tools model
#' @param ... other parameters
#' @rdname RunSCVI
#' @export
RunSCVI <- function(object, ...) {
    UseMethod("RunSCVI")
}

#' @rdname RunSCVI
#' @param object A seurat object
#' @return Return a Seurat object with the computed embedding
#'         with all the measure computed. For downstream Seurat analyses,
#'         use reduction='scvi'.
#' @export
RunSCVI.Seurat <- function(
        object,
        scvi_model = NULL,
        reduction.name = "scvi",
        reduction_key = "scvi_"
) {
    if (is.null(scvi_model)) {
        stop(paste0(
            "You have to pass a scvi model. The model can be"),
            "traine or not.")
    }

    scvi_model <- reticulate::r_to_py(scvi_model)
    if (!reticulate::py_to_r(scvi_model$is_trained)) {
        message("The model is not trained. Il will be trained now")
        scvi_model$train(max_epochs = max_epochs,
                    use_gpu = r_to_py(use_gpu),
                    early_stopping = r_to_py(early_stopping))
    }

    latent <- scvi_model$get_latent_representation()
    latent <- as.matrix(latent)
    rownames(latent) <- colnames(object)
    object[[reduction.name]] <-
    CreateDimReducObject(embeddings = latent,
                         key = reduction_key,
                         assay = DefaultAssay(object))
return(object)
}

